feat: Add concat operation to constraint streams (#297)
TimefoldAI/timefold-solver: docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc

See <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,`distinct()`>> for how to deal with duplicate tuples.
====

[[constraintStreamsConcat]]
==== Concat

Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.

> Unlike union of sets, concatenation of lists repeat duplicated elements.

Reviewer: ```suggestion
Unlike union of sets, concatenation of lists repeats duplicated elements.
```

If the two constraint stream parents share tuples, they will be repeated downstream.

> If this is undesired, use <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,distinct>>.

Reviewer: ```suggestion
If this is undesired, use the <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,`distinct` building block>>.
```

image::constraints-and-score/constraintStreamConcat.png[align="center"]

For example, to count the total number of cats and dogs:

[source,java,options="nowrap"]
----
    private Constraint countCatsAndDogs(ConstraintFactory constraintFactory) {
        return constraintFactory.forEach(Employee.class)
         .filter(employee -> employee.hasDog())
         .concat(constraintFactory.forEach(Employee.class)

>                      .filter(employee -> employee.hasCat()))

Reviewer: I am not sure this is a good example. The first question once a smart person sees this is: why not just use a much simple `.filter(hasCat || hasDog)`?

I think we need an example which doesn't have an obvious simpler faster solution.

Author: The obvious example would be to implement group by on an outer join, like this:

```java
factory.forEach(A.class)
    .join(B.class, ...)
    .concat(factory.forEach(A.class)
                               .ifNotExists(B.class, ...)
                               .expand(a -> null))
    .groupBy((a,b) -> a, conditionally((a,b) -> b != null, countBi())
    ...
```

Although the example would lead to the question "why no outer join" since that would simplify the code too

```java
factory.forEach(A.class)
    .outerJoin(B.class, ...)
    .groupBy((a,b) -> a, conditionally((a,b) -> b != null, countBi())
    ...
```

The other thing it can be used for is or'ing Joiners, which is unrelated to outer joins

```java
factory.forEach(A.class)
    .join(B.class, equal(a -> a.value, b -> b.value))
    .concat(factory.forEach(A.class)
                               .join(B.class, equal(a -> a.group, b -> b.group)))
    ...
```

Reviewer: If we're doing concat for the minimal constraint, why not show that? Maybe there's a simplified variant that isn't too much code while still showing the point?

I'm not worried if people start asking about outer join after they see this. First of all, we have a decent answer, this is faster because it doesn't need to do any sort of indexing. Second, if they do, and if there's enough of them, then maybe we eventually do it? It's an opportunity to coax some more real-world use cases out of people, and that's a good thing.

         .groupBy(ConstraintCollectors.count())
         .penalize(HardSoftScore.ONE_SOFT, count -> count)
         .asConstraint("Count cats and dogs");
    }
----

> This correctly count cats and dogs when an Employee has both a cat and a dog.

Reviewer: ```suggestion
This correctly counts cats and dogs when an Employee has both a cat and a dog.
```

> This correctly count cats and dogs when an Employee has both a cat and a dog.

Reviewer: ```suggestion
This correctly counts cats and dogs when an Employee has both a cat and a dog.
```

See <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,`distinct()`>> for how to deal with duplicate tuples.
====

[[constraintStreamsConcat]]
==== Concat

> Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.

Reviewer: ```suggestion
The `concat` building block allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
```

> If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.

Reviewer: ```suggestion
If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, `concat` acts like a concatenation of two lists.
```

Unlike union of sets, concatenation of lists repeats duplicated elements.

> If the two constraint stream parents share tuples, they will be repeated downstream.

Reviewer: ```suggestion
If the two constraint stream parents share tuples, which happens eg. when the streams being concatenated come from the same source of data, the tuples will be repeated downstream.
```

image::constraints-and-score/constraintStreamConcat.png[align="center"]

For example, to ensure each employee has a minimum number of assigned shifts:

[source,java,options="nowrap"]
----
    private Constraint ensureEachEmployeeHasAtLeastTwoShifts(ConstraintFactory constraintFactory) {
        return constraintFactory.forEach(Employee.class)
         .join(Shift.class, equal(Function.identity(), Shift::getEmployee))
         .concat(
             constraintFactory.forEach(Employee.class)
               .ifNotExists(Shift.class, equal(Function.identity(), Shift::getEmployee))
               .expand(employee -> (Shift) null)
         )
         .groupBy((employee, shift) -> employee,
                   conditionally((employee, shift) -> shift != null,
                                 countBi())
         )
         .filter((employee, shiftCount) -> shiftCount < employee.minimumAssignedShifts)
         .penalize(HardSoftScore.ONE_SOFT, (employee, shiftCount) -> employee.minimumAssignedShifts - shiftCount)
         .asConstraint("Minimum number of assigned shifts");
    }
----

> This correctly count the number of shifts each Employee has, *even when the Employee has no shifts*.

Reviewer: ```suggestion
This correctly counts the number of shifts each Employee has, *even when the Employee has no shifts*.
```

> If it was implemented without concat like this:

Reviewer: ```suggestion
If it was implemented without `concat` like this:
```
